examples/vision/retinanet.py

```diff
@@ -0,0 +1,700 @@
+
+    def _get_anchors(self, feature_height, feature_width, level):
+        """Returns anchors for a given level"""
+        rx = tf.range(feature_width, dtype=tf.float32) + 0.5
+        ry = tf.range(feature_height, dtype=tf.float32) + 0.5
+        centers = tf.stack(tf.meshgrid(rx, ry), axis=-1) * self._strides[level - 3]
+        centers = tf.expand_dims(centers, axis=-2)
+        centers = tf.tile(centers, [1, 1, self._num_anchors, 1])
+        wh = tf.tile(
+            self._anchor_dims[level - 3], [feature_height, feature_width, 1, 1]
+        )
+        anchors = tf.concat([centers, wh], axis=-1)
+        return tf.reshape(
+            anchors, [feature_height * feature_width * self._num_anchors, 4]
+        )
+
+    def get_anchors(self, image_height, image_width):
+        """Returns anchors for all levels"""
+        anchors = [
+            self._get_anchors(
+                tf.math.ceil(image_height / 2 ** i),
+                tf.math.ceil(image_width / 2 ** i),
+                i,
+            )
+            for i in range(3, 8)
+        ]
+        return tf.concat(anchors, axis=0)
+
+
+"""
+## Preprocessing image and labels
+"""
+
+
+def random_flip_lr(image, boxes):
+    """Flips image and boxes horizontally with 50% chance"""
+    if tf.random.uniform(()) > 0.5:
+        image = tf.image.flip_left_right(image)
+        boxes = tf.stack(
+            [1 - boxes[:, 2], boxes[:, 1], 1 - boxes[:, 0], boxes[:, 3]], axis=-1
+        )
+    return image, boxes
+
+
+def resize_and_pad_image(
+    image, min_side=800.0, max_side=1333.0, jitter=[640, 1024], stride=128.0
+):
+    """Resizes and pads image while preserving aspect ratio.
+        1. Resizes images so that the shorter side is equal to `min_side`
+        2. If the longer side is greater than `max_side`, then resize the image
+        with longer side equal to `max_side`
+        3. Pad with zeros on right and bottom to make the image shape divisible by
+        `stride`
+    """
+    image_shape = tf.cast(tf.shape(image)[:2], dtype=tf.float32)
+    if jitter is not None:
+        min_side = tf.random.uniform((), jitter[0], jitter[1], dtype=tf.float32)
+    ratio = min_side / tf.reduce_min(image_shape)
+    if ratio * tf.reduce_max(image_shape) > max_side:
+        ratio = max_side / tf.reduce_max(image_shape)
+    image_shape = ratio * image_shape
+    image = tf.image.resize(image, tf.cast(image_shape, dtype=tf.int32))
+    padded_image_shape = tf.cast(
+        tf.math.ceil(image_shape / stride) * stride, dtype=tf.int32
+    )
+    image = tf.image.pad_to_bounding_box(
+        image, 0, 0, padded_image_shape[0], padded_image_shape[1]
+    )
+    return image, image_shape, ratio
+
+
+def preprocess_data(sample):
+    image = sample["image"]
+    bbox = swap_xy(sample["objects"]["bbox"])
+    class_id = tf.cast(sample["objects"]["label"], dtype=tf.int32)
+
+    image, bbox = random_flip_lr(image, bbox)
+    image, image_shape, _ = resize_and_pad_image(image)
```

Comment:
This is not the exact replica of RetinaNet, right?
From the paper, "We use horizontal image flipping as the only form of data augmentation unless otherwise noted"

Reply:
Yes along with the horizontal flipping, the paper also talks about
Scale jittering: "We show results for our RetinaNet-101-800 model, trained with scale jitter and for 1.5× longer than the same model . . . " 

And resizing the shorter side:  "Larger backbone networks yield higher accuracy, but also slower inference speeds. Likewise
for input image scale (defined by the shorter image side)."

I added the padding after resizing to make the resultant image shape divisible by the stride of the model  `128` in this case.

Reply:
Oh ok, sounds good